### PR TITLE
better flexibility with children and panels

### PR DESCRIPTION
**Pull request by sccolbert**
Ready for discussion.

This is the result of feedback indicating that the current structure of panel children (observable list) is too rigid. Some users want to restrict the API for child types or otherwise implement custom APIs for adding children to a widget.

This PR updates Widget with the concept that children *exist* but the methods for adding children to a widget must be defined by a subclass. Two panel subclasses are provided as a convenience: one is `AbstractPanel` which defines an index-based API for accessing children, it also provides default implementations of core messaging behaviors which are necessary for proper functioning of children; the second is `Panel` which is a concrete implementation of the abstract class which stores its children in an internal array and exposes an API for adding children.

The core `Widget` class now exposes the APIs necessary to implement widgets which support children *without* forcing inheritance from a particular panel class. For these cases however, such a widget becomes responsible for implementing the child messaging behavior provided by `AbstractPanel`.

ping @jasongrout, @blink1073, @dwillmer, @afshin, @ellisonbg

### CODE REVIEW

**Comment by blink1073** [comment]
Two notes: it is probably best to view `widget.ts` rather than look at the diff, and @sccolbert, are you intending to put `title` back in there?

File: src/widget.ts [hunk 1]
```diff
@@ -478,14 +459,66 @@ class Widget extends NodeWrapper implements IDisposable, IMessageHandler {
...
+   * #### Notes
+   * This method **must** be reimplemented by subclasses which support
+   * children, or undefined behavior will result.
+   *
+   * This method is called automatically as needed. It should not be
+   * invoked directly by user code.
+   *
+   * The default implementation of this method is a no-op.
+   */
+  protected removeChild(child: Widget): void { }
```
Comment (blink1073) [inline src/widget.ts]: I'm not clear on why this and `disposeChildren` are needed here.  They are re-implemented as abstract methods on `AbstractPanel`.

Comment (sccolbert) [inline src/widget.ts]: 2 Reasons:

1) This is called by the `dispose` method of `Widget`, and `Widget` has no knowledge of the panel types.
2) It's no long required to inherit from `Panel` or `AbstractPanel` in order to create widgets which support children. Those two classes are there as conveniences which suit a large variety of use cases, but they may not suit everything. For those rare super-custom cases, the user can inherit `Widget` and implement everything themselves.

Comment (ellisonbg) [inline src/widget.ts]: Just to clarify, can you still just have panels that have an ObservableList
of children?

On Sun, Dec 13, 2015 at 10:33 AM, S. Chris Colbert <notifications@github.com
> wrote:

> In src/widget.ts
> <https://github.com/phosphorjs/phosphor-widget/pull/17#discussion_r47448096>
> :
>
> > +    child._parent = this;
> > +  }
> > +
> > +  /**
> > +   * Remove the specified child from the widget.
> > +   *
> > +   * #### Notes
> > +   * This method **must** be reimplemented by subclasses which support
> > +   * children, or undefined behavior will result.
> > +   *
> > +   * This method is called automatically as needed. It should not be
> > +   * invoked directly by user code.
> > +   *
> > +   * The default implementation of this method is a no-op.
> > +   */
> > +  protected removeChild(child: Widget): void { }
>
> 2 Reasons:
>
> 1) This is called by the dispose method of Widget, and Widget has no
> knowledge of the panel types.
> 2) It's no long required to inherit from Panel or AbstractPanel in order
> to create widgets which support children. Those two classes are there as
> conveniences which suit a large variety of use cases, but they may not suit
> everything. For those rare super-custom cases, the user can inherit Widget
> and implement everything themselves.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phosphorjs/phosphor-widget/pull/17/files#r47448096>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by sccolbert** [comment]
Not planning on keeping `.title`. I think it's a leaky abstraction, particularly when panels under the new system can support children of different types. Current thinking is to create a `TabItem` class which will uses to populate a `TabBar` or `TabPanel`. Something like:

``` typescript
class TabItem {
  text: string;
  closable: boolean;
  widget: Widget;
  // ...
}
```

File: src/widget.ts (hunk 1 above)
Comment (sccolbert) [inline src/widget.ts]: We could make one. But IMHO its no simpler to use than `Panel` proposed here. What use cases do you have in mind for that?

One thing I didnt like about using observable list for children is that the list must be kept unique (no duplicate children), so it doesn't behave *purely* like a normal list, and all multi-item operations need to be decomposed into multiple single-item equivalents.

**Comment by blink1073** [comment]
Okay, :+1: from me.

**Comment by sccolbert** [comment]
Ready to merge.

**Comment by sccolbert** [comment]
I'm going to merge and release so I can keep moving forward. If there's a large outcry, we can rewind.

**Pull request merged**